QWeb/internal/frame.py

```diff
@@ -207,7 +208,63 @@ def search_from_frames(driver=None, current_frame=None):
                 return web_element
             driver.switch_to.default_content()
             raise QWebTimeoutError('From frame decorator: Unable to locate element in given time')
-        return search_from_frames()
+
+        # pylint: disable=W0102
+        def search_from_frames_safari(driver=None, current_frame=None, parent_tree=[]):
+            keep_frame = kwargs.get('stay_in_current_frame', CONFIG['StayInCurrentFrame'])
+            if keep_frame:
+                return fn(*args, **kwargs)
+            err = None
+            if not driver:
+                driver = browser.get_current_browser()
+                driver.switch_to.default_content()
+            if current_frame is not None:
+                try:
+                    driver.switch_to.frame(current_frame)
+                    logger.debug('switching to childframe {}'.format(str(fn)))
+                except (StaleElementReferenceException, WebDriverException) as e:
+                    logger.warn(e)
+                    driver.switch_to.default_content()
+                    raise e
+            try:
+                web_element = fn(*args, **kwargs)
+            except QWebElementNotFoundError as e:
+                err = e
+                web_element = None
+            if is_valid(web_element):
+                return web_element
+            start = time.time()
+            timeout = CONFIG['FrameTimeout']
+            while time.time() < timeout + start:
+                frames = fc.check_frames(driver)
```

Comment:
It is safari specific function with few differences. Was left there in purpose. We first had only one and did the same for all browsers, but since switch_to.parent_frame() works correctly on all other browsers we decided to keep the old, working one for other browsers and introduce a safari specific solution only for safari.

Since it's recursive and we need to keep track on traversed frames in safari (parent_tree), we think it's easier to remember 6 months from now when there's strictly safari specific funtion.